ruma-appservice-api/src/event/push_events/v1.rs

```diff
@@ -41,6 +41,84 @@ impl IncomingRequest {
+    /// instead of the original `push_events::IncomingRequest`. It puts all events with a `room_id`
+    /// into the `JoinedRoom`'s `timeline` or into `ephemeral` if no `state_key` and `event_id` is
+    /// set. The rationale behind that is that incoming Appservice transactions from the homeserver
+    /// are not necessarily bound to a specific user but can cover a multitude of namespaces, and
+    /// as such the Appservice basically only "observes joined rooms".
+    ///
+    /// Additionally events with type `m.presence` are put into the response's `presence`
+    /// accordingly.
+    #[cfg(feature = "helper")]
+    pub fn try_into_sync_response(
+        self,
+        next_batch: impl Into<String>,
+    ) -> Result<ruma_client_api::r0::sync::sync_events::Response, serde_json::Error> {
+        use ruma_client_api::r0::sync::sync_events;
+        use ruma_events::{presence::PresenceEvent, AnySyncEphemeralRoomEvent, AnySyncRoomEvent};
+        use ruma_identifiers::RoomId;
+        use serde::{de::IgnoredAny, Deserialize};
+
+        #[derive(Debug, Deserialize)]
+        struct EventDeHelper {
+            /// the Matrix event type string "m.room.whatever".
+            #[serde(rename = "type")]
+            ev_type: String,
+
+            /// If `state_key` is present the event will be deserialized as a state event.
+            state_key: Option<IgnoredAny>,
+
+            /// If no `state_key` is found but an `event_id` is present the event
+            /// will be deserialized as a message event.
+            event_id: Option<IgnoredAny>,
+
+            /// If no `event_id` or `state_key` are found but a `room_id` is present
+            /// the event will be deserialized as a ephemeral event.
+            room_id: Option<RoomId>,
+        }
+
+        let mut response = sync_events::Response::new(next_batch.into());
+
+        for raw_event in self.events {
+            let helper: EventDeHelper = serde_json::from_str(raw_event.json().get())?;
+            let event_json = Raw::into_json(raw_event);
+
+            match helper.room_id {
+                Some(room_id) => {
+                    let join = response
+                        .rooms
+                        .join
+                        .entry(room_id)
+                        .or_insert_with(sync_events::JoinedRoom::new);
+
+                    if helper.state_key.is_some() || helper.event_id.is_some() {
+                        join.timeline.events.push(Raw::<AnySyncRoomEvent>::from_json(event_json))
+                    } else {
+                        join.ephemeral
+                            .events
+                            .push(Raw::<AnySyncEphemeralRoomEvent>::from_json(event_json));
+                    }
+                }
+                None =>
+                {
+                    #[allow(clippy::single_match)]
+                    match helper.ev_type.as_ref() {
+                        "m.presence" => response
```

Comment:
After looking into it a bit more, it seems that appservices currently only get PDUs pushed, not EDUs, since [MSC2409](https://github.com/matrix-org/matrix-doc/pull/2409) exists. So I guess the no-room_id arm is not necessary.

Reply:
Okay so really the current handling presence and ephemeral room events can be removed since they won't be part of `self.events` when they are introduced?

Reply:
I think that means this whole method will be line 10 lines of code, huh? :D

Reply:
> Okay so really the current handling presence and ephemeral room events can be removed since they won't be part of self.events when they are introduced?

All EDUs according to that MSC, yeah

> I think that means this whole method will be line 10 lines of code, huh? :D

Yep, thanks for all the feedback & reviews so far, learned a lot along the way :D